crates/jit/src/link.rs

```diff
@@ -19,18 +19,27 @@ use wasmtime_runtime::VMFunctionBody;
 /// TODO refactor logic to remove panics and add defensive code the image data
 /// becomes untrusted.
 pub fn link_module(
-    obj: &File,
+    obj_data: &mut [u8],
     module: &Module,
-    code_range: &mut [u8],
     finished_functions: &PrimaryMap<DefinedFuncIndex, *mut [VMFunctionBody]>,
-) {
-    // Read the ".text" section and process its relocations.
-    let text_section = obj.section_by_name(".text").unwrap();
-    let body = code_range.as_ptr() as *const VMFunctionBody;
+) -> Result<(), String> {
+    let obj = File::parse(obj_data).map_err(|_| "Unable to read obj".to_string())?;
+
+    for section in obj.sections() {
+        let body_offset = match section.file_range() {
+            Some((start, _)) => start as usize,
+            None => {
+                continue;
+            }
+        };
+        let body = unsafe { obj_data.as_ptr().add(body_offset) } as *const VMFunctionBody;
```

Comment:
Could the `unsafe` be avoided here by reading the second data directly from the `section` from the iterator?